### PR TITLE
Add PMTCT optout utils and functionality

### PR DESCRIPTION
**Pull request by NicolaasDieBaas**

### CODE REVIEW

**Comment by NicolaasDieBaas** [comment]
@gsvr please review.  You'll notice that hub.change_registration currently performs a 'PATCH' request... this we'll fix to be 'POST' when we update seed-jsbox-utils on npm

**Comment by NicolaasDieBaas** [comment]
@gsvr please review

File: src/ussd_pmtct.js [hunk 1]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
             return subscriptionTypeMapping[messageset_id];
         };
 
+        self.getSubscriptionsByMsisdn = function(im, msisdn) {
```
Comment (gsvr) [inline src/ussd_pmtct.js]: Vumi

File: src/ussd_pmtct.js [hunk 2]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
...
+                    params: {
+                        "query": "toaddr=" + msisdn
+                    }
+                })
+                .then(function(result) {
+                    return result.data;
+                });
+        };
+
+        self.deactivateVumiSubscriptions = function(im, contact) {
```
Comment (gsvr) [inline src/ussd_pmtct.js]: pass msisdn, not contact

File: src/ussd_pmtct.js [hunk 3]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
...
+
+            var http = new JsonApi(im, {
+                headers: {
+                    'Authorization': ['ApiKey ' + username + ':' + api_key]
+                }
+            });
+            var msisdn = Object.keys(contact.details.addresses.msisdn)[0];
+            return self
+                .getSubscriptionsByMsisdn(im, msisdn)
+                .then(function(update) {
```
Comment (gsvr) [inline src/ussd_pmtct.js]: call this subscriptions rather than update

File: src/ussd_pmtct.js [hunk 4]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
...
+                        return http.patch(subscription_base_url + endpoint, {
+                                data: update
+                            });
+                    } else {
+                        return Q();
+                    }
+                });
+        };
+
+        self.subscribeVumiContactToLossMessages = function(im, identity) {
```
Comment (gsvr) [inline src/ussd_pmtct.js]: remove this function - can call `postVumiLossSubscription` directly

File: src/ussd_pmtct.js [hunk 5]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
...
+                "babyloss": "8"
+            };
+            var username = im.config.vumi.username;
+            var api_key = im.config.vumi.api_key;
+
+            var subscription_base_url = im.config.vumi.subscription_url;
+            var endpoint = "subscription/";
+
+            var sub_info = {
+                // contact_key: contact.key, ???
```
Comment (gsvr) [inline src/ussd_pmtct.js]: add this, yes.
Also please update the identity update to save the contact_key with the user_account.

File: src/ussd_pmtct.js [hunk 6]
```diff
@@ -231,7 +353,7 @@ go.app = function() {
                         return self.im.user
                         .set_lang(self.im.user.answers.contact_identity.details.lang_code || "eng_ZA")
                         .then(function(lang_set_response) {
-                            return self.states.create("state_optout_reason_menu");
+                            return self.states.create("state_optout_reason_menu", identity);
```
Comment (gsvr) [inline src/ussd_pmtct.js]: please save it as a user answer and reuse that rather than passing it through everywhere.

File: src/ussd_pmtct.js [hunk 7]
```diff
@@ -153,6 +159,122 @@ go.app = function() {
...
+            var subscription_base_url = im.config.vumi.subscription_url;
+            var endpoint = "subscription/";
+
+            var sub_info = {
+                // contact_key: contact.key, ???
+                lang: im.user.lang,
+                message_set: "/api/v1/message_set/" + optoutReasonToSubTypeMapping[im.user.answers.optout_reason] + "/",
+                next_sequence_number: 1,
+                schedule: "/api/v1/periodic_task/3/",  // 3 = twice per week
+                to_addr: Object.keys(contact.details.addresses.msisdn)[0],
```
Comment (gsvr) [inline src/ussd_pmtct.js]: this is in `im.user.answers.msisdn`

**Comment by NicolaasDieBaas** [comment]
@gsvr second pass review please; made the small changes you highlighted before.

**Comment by NicolaasDieBaas** [comment]
@gsvr is this ok to land now?

**Comment by gsvr** [comment]
:+1: \o/

**Pull request merged**